### PR TITLE
Handle situations where PeerInit and PierceFirewall msgs have no token

### PR DESCRIPTION
**Pull request by mathiascode**

### CODE REVIEW

**Comment by kiplingw** [comment]
Thanks @mathiascode. Fixing various network i/o and NAT issues people sometimes have should be a priority over new features. Glad to see you're working on this.

**Comment by mathiascode** [comment]
That's true, although sometimes you have to focus on other things to avoid burning yourself out on these network issues. Hopefully we can improve network compatibility further. :)

**Comment by kiplingw** [comment]
Yup. Try to find friends who are having an issue downloading from each other, and then try and look further at what went on. The benefit of relying on friends is you have access to both peers' systems.

File: pynicotine/pynicotine.py [hunk 1]
```diff
@@ -628,23 +628,28 @@ def out_conn(self, msg):
...
+        # Sometimes a token is seemingly not sent by the peer, check if the
+        # IP address matches in that case
         token = msg.token
 
         for i in self.peerconns:
-            if i.token == token and i.conn is None:
-                conn = msg.conn.conn
+            if i.conn is None:
+                if token is not None and i.token == token or \
+                        token is None and i.token is not None and i.addr == msg.conn.addr:
```
Comment (hboetes) [inline pynicotine/pynicotine.py:639]: That's hard to read. Would a set of parens make it easier?

**Pull request merged**